immuclient crashes when starting audit-mode manually

immuclient crashes when starting audit-mode manually (despite a running service)

./immuclient audit-mode
Using config file: /etc/immudb/immuclient.toml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xb21c52]
goroutine 1 [running]:
github.com/codenotary/immudb/cmd/immuclient/audit.(*executable).Run(0xc0003c6f20)
        /source/cmd/immuclient/audit/executable.go:75 +0x52
github.com/takama/daemon.(*systemDRecord).Run(0xc00035c5a0, 0xec9e80, 0xc0003c6f20, 0xebb580, 0xc000466070, 0xc0003c6f80, 0x0)
        /go/pkg/mod/github.com/codenotary/daemon@v0.0.0-20200507161650-3d4bcb5230f4/daemon_linux_systemd.go:206 +0xa1
github.com/codenotary/immudb/cmd/immuclient/audit.(*auditAgent).Manage(0xc0003da000, 0x168cea8, 0x0, 0x0, 0xc000386c00, 0xc000157cf0, 0x1, 0xd9b332)
        /source/cmd/immuclient/audit/auditagent.go:184 +0x1094
github.com/codenotary/immudb/cmd/immuclient/audit.Init(0x168cea8, 0x0, 0x0)
        /source/cmd/immuclient/audit/init.go:33 +0x9e
github.com/codenotary/immudb/cmd/immuclient/command.(*commandline).auditmode.func1(0xc0003bc580, 0x168cea8, 0x0, 0x0, 0x0, 0x0)
        /source/cmd/immuclient/command/misc.go:90 +0x3f
github.com/spf13/cobra.main.main()
        /source/cmd/immuclient/immuclient.go:35 +0xce

--- a/Makefile
+++ b/Makefile
@@ -97,6 +97,7 @@ immutest-static:
 vendor:
 	$(GO) mod vendor
 
+# To view coverage as HTML run: go tool cover -html=coverage.txt
 .PHONY: test
 test:
 	$(GO) vet ./...

--- a/cmd/immuclient/audit/auditagent.go
+++ b/cmd/immuclient/audit/auditagent.go
@@ -17,7 +17,6 @@ limitations under the License.
 package audit
 
 import (
-	"encoding/json"
 	"fmt"
 	"io"
 	"os"
@@ -25,7 +24,6 @@ import (
 	"github.com/codenotary/immudb/pkg/server"
 
 	"github.com/codenotary/immudb/cmd/immuclient/service"
-	"github.com/codenotary/immudb/pkg/api/schema"
 	"github.com/codenotary/immudb/pkg/client"
 	"github.com/codenotary/immudb/pkg/client/auditor"
 	"github.com/codenotary/immudb/pkg/logger"
@@ -41,31 +39,14 @@ type AuditAgent interface {
 type auditAgent struct {
 	daemon.Daemon
 	cycleFrequency int
-	promot         promotheusExporter
+	metrics        prometheusMetrics
 	ImmuAudit      auditor.Auditor
 	immuc          client.ImmuClient
-	rootStorage    io.Writer
 	firstRun       bool
 	opts           *client.Options
 	logger         logger.Logger
 	Pid            server.PIDFile
 	logfile        io.Writer
-}
-
-func (a *auditAgent) writeRoot(root *schema.Root, t string) error {
-	m := make(map[string]interface{}, 2)
-	m["root"] = root
-	m["timeStamp"] = t
-	data, err := json.Marshal(m)
-	if err != nil {
-		return err
-	}
-	data = append(data, []byte("\n")...)
-	_, err = a.rootStorage.Write(data)
-	if err != nil {
-		return err
-	}
-	return nil
 }
 
 func (a *auditAgent) Manage(args []string) (string, error) {
@@ -80,8 +61,7 @@ func (a *auditAgent) Manage(args []string) (string, error) {
 	exec := newExecutable(a)
 
 	if command == "install" {
-		_, err := a.InitAgent()
-		if err != nil {
+		if _, err = a.InitAgent(); err != nil {
 			QuitToStdErr(err.Error())
 		}
 		localFile = viper.GetString("local-file")
@@ -210,16 +190,4 @@ func options() *client.Options {
 			WithClientCAs(clientcas)
 	}
 	return options
-}
-
-func getCommands() string {
-	cmds := []string{"install", "uninstall", "start", "stop", "status", "restart"}
-	for i := range os.Args {
-		for j := range cmds {
-			if os.Args[i] == cmds[j] {
-				return cmds[j]
-			}
-		}
-	}
-	return ""
 }

--- a/cmd/immuclient/audit/auditagent.go
+++ b/cmd/immuclient/audit/auditagent.go
@@ -19,9 +19,10 @@ package audit
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/codenotary/immudb/pkg/server"
 	"io"
 	"os"
+
+	"github.com/codenotary/immudb/pkg/server"
 
 	"github.com/codenotary/immudb/cmd/immuclient/service"
 	"github.com/codenotary/immudb/pkg/api/schema"
@@ -180,7 +181,9 @@ func (a *auditAgent) Manage(args []string) (string, error) {
 	}
 	a.logfile = logfile
 	a.logger = logger.NewSimpleLogger("immuclientd", logfile)
-	a.InitAgent()
+	if _, err := a.InitAgent(); err != nil {
+		return "", err
+	}
 	return a.Run(exec)
 }
 

--- a/cmd/immuclient/audit/auditor.go
+++ b/cmd/immuclient/audit/auditor.go
@@ -20,12 +20,12 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"github.com/codenotary/immudb/pkg/server"
 	"os"
 	"path/filepath"
-	"strconv"
 	"strings"
 	"time"
+
+	"github.com/codenotary/immudb/pkg/server"
 
 	"github.com/codenotary/immudb/pkg/client"
 	"github.com/codenotary/immudb/pkg/client/auditor"
@@ -46,55 +46,36 @@ func (cAgent *auditAgent) InitAgent() (AuditAgent, error) {
 		return nil, fmt.Errorf("Initialization failed: %s \n", err.Error())
 	}
 	ctx := context.Background()
-	
+
 	pidPath := viper.GetString("pidfile")
-	if pidPath != ""{
+	if pidPath != "" {
 		if cAgent.Pid, err = server.NewPid(pidPath); err != nil {
 			cAgent.logger.Errorf("failed to write pidfile: %s", err)
 			return nil, err
 		}
 	}
 
-	freqstr := os.Getenv("audit-agent-interval")
 	cAgent.cycleFrequency = 60
+	if freqstr := os.Getenv("audit-agent-interval"); freqstr != "" {
+		d, err := time.ParseDuration(freqstr)
+		if err != nil {
+			return nil, err
+		}
+		cAgent.cycleFrequency = int(d.Seconds())
+	}
 	sclient := cAgent.immuc.GetServiceClient()
 	serverID, err := client.GetServerUuid(ctx, *sclient)
 	if serverID == "" || err != nil {
 		serverID = "unknown"
 	}
 	cAgent.promot.init(serverID)
-	if freqstr != "" {
-		s := freqstr[len(freqstr)-1]
-		switch string(s) {
-		case "s":
-			freq, err := strconv.Atoi(freqstr[:len(freqstr)-1])
-			if err == nil && freq > 60 {
-				cAgent.cycleFrequency = freq
-			}
-		case "m":
-			freq, err := strconv.Atoi(freqstr[:len(freqstr)-1])
-			if err == nil && freq > 1 {
-				cAgent.cycleFrequency = freq * 60
-			}
-		case "h":
-			freq, err := strconv.Atoi(freqstr[:len(freqstr)-1])
-			if err == nil && freq >= 1 {
-				cAgent.cycleFrequency = (freq * 60 * 60)
-			}
-		default:
-			freq, err := strconv.Atoi(freqstr[:len(freqstr)-1])
-			if err == nil && freq >= 1 {
-				cAgent.cycleFrequency = freq
-			}
-		}
-	}
-
 	cliOpts := cAgent.immuc.GetOptions()
 	ctx = context.Background()
-	if viper.GetString("audit-username") != "" || viper.GetString("audit-password") != "" {
-		_, err := cAgent.immuc.Login(ctx, []byte(viper.GetString("audit-username")), []byte(viper.GetString("audit-password")))
-		if err != nil {
-			return nil, errors.New("Invalid login operation")
+	auditUsername := []byte(viper.GetString("audit-username"))
+	auditPassword := []byte(viper.GetString("audit-password"))
+	if len(auditUsername) > 0 || len(auditPassword) > 0 {
+		if _, err = cAgent.immuc.Login(ctx, auditUsername, auditPassword); err != nil {
+			return nil, fmt.Errorf("Invalid login operation: %v", err)
 		}
 	}
 

--- a/cmd/immuclient/audit/auditor.go
+++ b/cmd/immuclient/audit/auditor.go
@@ -22,7 +22,6 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
-	"strings"
 	"time"
 
 	"github.com/codenotary/immudb/pkg/server"
@@ -68,7 +67,7 @@ func (cAgent *auditAgent) InitAgent() (AuditAgent, error) {
 	if serverID == "" || err != nil {
 		serverID = "unknown"
 	}
-	cAgent.promot.init(serverID)
+	cAgent.metrics.init(serverID)
 	cliOpts := cAgent.immuc.GetOptions()
 	ctx = context.Background()
 	auditUsername := []byte(viper.GetString("audit-username"))
@@ -85,16 +84,9 @@ func (cAgent *auditAgent) InitAgent() (AuditAgent, error) {
 		viper.GetString("audit-username"),
 		viper.GetString("audit-password"),
 		cache.NewHistoryFileCache(filepath.Join(os.TempDir(), "auditor")),
-		cAgent.promot.exporter, cAgent.logfile)
+		cAgent.metrics.updateMetrics, cAgent.logfile)
 	if err != nil {
 		return nil, err
 	}
 	return cAgent, nil
-}
-
-func filename() string {
-	t := time.Now().String()[:16]
-	t = strings.ReplaceAll(t, " ", "_")
-	t = fmt.Sprintf("audit_roots_%s.txt", t)
-	return t
 }

--- a/cmd/immuclient/audit/executable.go
+++ b/cmd/immuclient/audit/executable.go
@@ -38,9 +38,9 @@ func newExecutable(a *auditAgent) *executable {
 func (e *executable) Start() {
 	go func() {
 		go func() {
-			e.a.promot.port = viper.GetString("prometheus-port")
-			e.a.promot.port = viper.GetString("prometheus-host")
-			err := e.a.promot.startPromoExporter()
+			e.a.metrics.port = viper.GetString("prometheus-port")
+			e.a.metrics.port = viper.GetString("prometheus-host")
+			err := e.a.metrics.startServer()
 			if err != nil {
 				fmt.Println(err.Error())
 			}
@@ -56,9 +56,9 @@ func (e *executable) Stop() {
 
 func (e *executable) Run() {
 	go func() {
-		e.a.promot.port = viper.GetString("prometheus-port")
-		e.a.promot.address = viper.GetString("prometheus-host")
-		err := e.a.promot.startPromoExporter()
+		e.a.metrics.port = viper.GetString("prometheus-port")
+		e.a.metrics.address = viper.GetString("prometheus-host")
+		err := e.a.metrics.startServer()
 		if err != nil {
 			fmt.Println(err.Error())
 		}

--- a/cmd/immuclient/audit/metrics.go
+++ b/cmd/immuclient/audit/metrics.go
@@ -26,7 +26,7 @@ import (
 	"github.com/spf13/viper"
 )
 
-type promotheusExporter struct {
+type prometheusMetrics struct {
 	port           string
 	address        string
 	server_address string
@@ -54,7 +54,7 @@ var (
 	)
 )
 
-func (p *promotheusExporter) init(serverid string) {
+func (p *prometheusMetrics) init(serverid string) {
 	p.server_address = fmt.Sprintf("%s:%s", viper.GetString("immudb-address"), viper.GetString("immudb-port"))
 	p.server_id = serverid
 	prometheus.MustRegister(AuditResultPerServer, AuditCurrRootPerServer, AuditRunAtPerServer, AuditPrevRootPerServer)
@@ -64,7 +64,7 @@ func (p *promotheusExporter) init(serverid string) {
 	AuditPrevRootPerServer.WithLabelValues(p.server_id, p.server_address).Set(-1)
 }
 
-func (p *promotheusExporter) startPromoExporter() error {
+func (p *prometheusMetrics) startServer() error {
 
 	http.Handle("/", promhttp.Handler())
 	fmt.Printf("Beginning to serve on port %s:%s \n", p.address, p.port)
@@ -87,7 +87,7 @@ func newAuditGaugeVec(name string, help string) *prometheus.GaugeVec {
 	)
 }
 
-func (p *promotheusExporter) exporter(
+func (p *prometheusMetrics) updateMetrics(
 	serverID string,
 	serverAddress string,
 	checked bool,

